### ISSUE TITLE
Type narrowing for `LiteralString` does not work

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#3466]
**Describe the bug**
Given the following code:

```python
from collections.abc import Sequence
from typing import Any, TypeGuard
from typing_extensions import LiteralString, reveal_type


def _is_literal_string(obj: Any) -> TypeGuard[LiteralString]:
    return isinstance(obj, str)


def foo(s: str | Sequence[str]) -> None:
    if isinstance(s, str):
        reveal_type(s)
    else:
        reveal_type(s)


def bar(s: LiteralString | Sequence[LiteralString]) -> None:
    if _is_literal_string(s):
        reveal_type(s)
    else:
        reveal_type(s)
```

The output on the command line is as follows:

```
test.py:8:21 - information: Type of "s" is "str"
test.py:10:21 - information: Type of "s" is "Sequence[str]"
test.py:19:21 - information: Type of "s" is "LiteralString"
test.py:21:21 - information: Type of "s" is "LiteralString | Sequence[LiteralString]"
```

**Expected behavior**
The type reported on line 21 should be `Sequence[LiteralString]`.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line

Which version?
1.1.247

**Comment by erictraut** [comment]
The current behavior is correct. `TypeGuard` does not perform type narrowing in the negative ("else") case. This was intentional and is specified in [PEP 647](https://peps.python.org/pep-0647/#typeguard-type).

There has been discussion of a variant of `TypeGuard` that would be more restrictive (and therefore less flexible) but would allow for type narrowing in the negative case. I've proposed that we call this `StrictTypeGuard`, and it is prototyped in pyright. It has no runtime support currently, so you won't be able to use it in your code unless you use some hackery.

```python
if TYPE_CHECKING:
    from typing_extensions import StrictTypeGuard
else:
    StrictTypeGuard = Any

def _is_literal_string(obj: Any) -> StrictTypeGuard[LiteralString]:
    return isinstance(obj, str)
```

**Comment by erictraut** [comment]
For more details about the `StrictTypeGuard` proposal, see [this discussion](https://mail.python.org/archives/list/typing-sig@python.org/message/JTO3WKRKQFW5YFP3ZSMNTRIEA5NPJEM6/).

**Comment by bryanforbes** [comment]
Thanks, @erictraut. That's exactly what I was looking for. Hopefully `StrictTypeGuard` gains traction.
